src/main/java/com/researchspace/api/v1/controller/GalleryFilestoresApiController.java

```diff
@@ -0,0 +1,217 @@
+import java.util.List;
+import javax.servlet.ServletOutputStream;
+import javax.servlet.http.HttpServletResponse;
+import javax.validation.Valid;
+import javax.ws.rs.NotFoundException;
+import lombok.NoArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang.StringUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.validation.BindException;
+import org.springframework.validation.BindingResult;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestAttribute;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestParam;
+
+@Slf4j
+@NoArgsConstructor
+@ApiController
+public class GalleryFilestoresApiController extends GalleryFilestoresBaseApiController
+    implements GalleryFilestoresApi {
+
+  @Autowired private NfsManager nfsManager;
+  @Autowired private NfsFileHandler nfsFileHandler;
+
+  @Override
+  public List<NfsFileStoreInfo> getUserFilestores(@RequestAttribute(name = "user") User user) {
+    assertFilestoresApiEnabled(user);
+    return nfsManager.getFileStoreInfosForUser(user);
+  }
+
+  @Override
+  public ApiNfsRemotePathBrowseResult browseFilestore(
+      @PathVariable Long filestoreId,
+      @RequestParam(value = "remotePath", required = false) String browsePath,
+      @RequestAttribute(name = "user") User user)
+      throws IOException {
+
+    assertFilestoresApiEnabled(user);
+    NfsFileStore filestore = nfsManager.getNfsFileStore(filestoreId);
+    NfsFileSystem filesystem = filestore.getFileSystem();
+    NfsClient nfsClient = credentialsStore.getNfsClientWithStoredCredentials(user, filesystem);
+
+    String combinedPath = filestore.getPath();
+    if (StringUtils.isNotEmpty(browsePath)) {
+      combinedPath += browsePath;
+    }
+
+    NfsFileTreeNode fileTree = nfsClient.createFileTree(combinedPath, null, filestore);
+    return getRemotePathBrowseResult(filesystem, nfsClient, fileTree);
+  }
+
+  private ApiNfsRemotePathBrowseResult getRemotePathBrowseResult(
+      NfsFileSystem filesystem, NfsClient nfsClient, NfsFileTreeNode fileTree) {
+
+    return new ApiNfsRemotePathBrowseResult(
+        fileTree,
+        filesystem.toFileSystemInfo(),
+        nfsClient.supportsExtraDirs() && !filesystem.fileSystemRequiresUserRootDirs(),
+        nfsClient.supportsCurrentDir(),
+        nfsClient.getUsername());
+  }
+
+  @Override
+  public void downloadFromFilestore(
+      @PathVariable Long filestoreId,
+      @RequestParam(value = "remotePath", required = false) String remotePath,
+      @RequestParam(name = "remoteId", required = false) Long remoteId,
+      @RequestAttribute(name = "user") User user,
+      HttpServletResponse response)
+      throws IOException {
+
+    assertFilestoresApiEnabled(user);
+    if (StringUtils.isEmpty(remotePath) && remoteId == null) {
```

Comment:
if you use `isBlank` you can check also if there are only special character like TAB of NewLine